### PR TITLE
pybind/rados: add Ioctx::aio_setxattr() and remove_omap_range2()

### PR DESCRIPTION
**Pull request by zjcmszh**
Signed-off-by: Zhang Jiao <zhangjiao_yewu@cmss.chinamobile.com>

### CODE REVIEW

**Comment by zjcmszh** [comment]
@tchaikov updated，please check

File: src/test/pybind/test_rados.py [hunk 1]
```diff
@@ -578,6 +578,26 @@ def test_clear_omap(self):
...
             eq(list(iter), [])
 
+    def test_remove_omap_ramge2(self):
+        keys = ("1", "2", "3", "4")
+        values = (b"a", b"bb", b"ccc", b"dddd")
+        with WriteOpCtx() as write_op:
+            self.ioctx.set_omap(write_op, keys, values)
+            self.ioctx.operate_write_op(write_op, "test_obj")
+        with ReadOpCtx() as read_op:
+            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op,("1","2","3","4"))
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py:588]: please use `keys` here instead of repeating its value.

File: src/test/pybind/test_rados.py [hunk 2]
```diff
@@ -578,6 +578,26 @@ def test_clear_omap(self):
...
+            eq(ret, 0)
+            self.ioctx.operate_read_op(read_op, "test_obj")
+            eq(list(iter), [("1", b"a"), ("2", b"bb"), ("3", b"ccc"),("4", b"dddd")])
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py:591]: please use `zip(keys, values)`.

File: src/test/pybind/test_rados.py [hunk 3]
```diff
@@ -578,6 +578,26 @@ def test_clear_omap(self):
...
+        with WriteOpCtx() as write_op:
+            self.ioctx.remove_omap_range2(write_op, "1", "4")
+            self.ioctx.operate_write_op(write_op, "test_obj")
+        with ReadOpCtx() as read_op:
+            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op,("1","2","3","4"))
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py:596]: use `keys`

File: src/test/pybind/test_rados.py (hunk 2 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: thank you very much for your kindly advice, updated.

File: src/test/pybind/test_rados.py (hunk 1 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: updated, please check

File: src/test/pybind/test_rados.py (hunk 3 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: updated

**Review by tchaikov — APPROVED** [review]

File: src/test/pybind/test_rados.py [hunk 4]
```diff
@@ -578,6 +578,26 @@ def test_clear_omap(self):
...
+        keys = ("1", "2", "3", "4")
+        values = (b"a", b"bb", b"ccc", b"dddd")
+        with WriteOpCtx() as write_op:
+            self.ioctx.set_omap(write_op, keys, values)
+            self.ioctx.operate_write_op(write_op, "test_obj")
+        with ReadOpCtx() as read_op:
+            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, keys)
+            eq(ret, 0)
+            self.ioctx.operate_read_op(read_op, "test_obj")
+            eq(list(iter), zip(keys, values))
```
Comment (tchaikov) [inline src/test/pybind/test_rados.py:591]:
```
2020-11-14T10:40:57.076 INFO:tasks.workunit.client.0.smithi031.stderr:======================================================================
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:FAIL: test_rados.TestIoctx.test_remove_omap_ramge2
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:----------------------------------------------------------------------
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:Traceback (most recent call last):
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:  File "/usr/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:    self.test(*self.arg)
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:  File "/home/ubuntu/cephtest/clone.client.0/src/test/pybind/test_rados.py", line 591, in test_remove_omap_ramge2
2020-11-14T10:40:57.077 INFO:tasks.workunit.client.0.smithi031.stderr:    eq(list(iter), zip(keys, values))
2020-11-14T10:40:57.078 INFO:tasks.workunit.client.0.smithi031.stderr:AssertionError: [('1', b'a'), ('2', b'bb'), ('3', b'ccc'), ('4', b'dddd')] != <zip object at 0x7f880ccbea08>
```

could you please run the test locally?

**Review by tchaikov — CHANGES_REQUESTED** [review]

**Comment by tchaikov** [comment]
@zjcmszh ping?

File: src/test/pybind/test_rados.py (hunk 4 above)
Comment (zjcmszh) [reply src/test/pybind/test_rados.py:591]: @tchaikov Too busy these days, I am sorry reply late.

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: test code:

`import rbd
from rados import (Rados, WriteOpCtx, ReadOpCtx)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool_01')
        assert self.rados.pool_exists('pool_01')
        self.ioctx = self.rados.open_ioctx('pool_01')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool_01')
        self.rados.shutdown()

    def test_remove_omap_ramge2(self):
        keys = ("1", "2", "3", "4")
        values = (b"a", b"bb", b"ccc", b"dddd")
        with WriteOpCtx() as write_op:
            self.ioctx.set_omap(write_op, keys, values)
            self.ioctx.operate_write_op(write_op, "test_obj")
        with ReadOpCtx() as read_op:
            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, keys)
            eq(ret, 0)
            self.ioctx.operate_read_op(read_op, "test_obj")
            eq(list(iter), list(zip(keys, values)))
        with WriteOpCtx() as write_op:
            self.ioctx.remove_omap_range2(write_op, "1", "4")
            self.ioctx.operate_write_op(write_op, "test_obj")
        with ReadOpCtx() as read_op:
            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, keys)
            eq(ret, 0)
            self.ioctx.operate_read_op(read_op, "test_obj")
            eq(list(iter), [("4", b"dddd")])

if __name__ == '__main__':
    unittest.main()
`
test results:
 
`root@jw-KVM:/data/ceph/build# python3 test_range2.py 
.
----------------------------------------------------------------------
Ran 1 test in 2.684s

OK
`

Comment (zjcmszh) [reply src/test/pybind/test_rados.py]: test code:
```
import rbd
from rados import (Rados, WriteOpCtx, ReadOpCtx)
from nose.tools import eq_ as eq, ok_ as ok, assert_raises
import rados
import unittest

class TestIoctx(unittest.TestCase):
    def setUp(self):
        self.rados = Rados(conffile='./ceph.conf')
        self.rados.connect()
        self.rados.create_pool('pool_01')
        assert self.rados.pool_exists('pool_01')
        self.ioctx = self.rados.open_ioctx('pool_01')
    def tearDown(self):
        self.ioctx.close()
        self.rados.delete_pool('pool_01')
        self.rados.shutdown()

    def test_remove_omap_ramge2(self):
        keys = ("1", "2", "3", "4")
        values = (b"a", b"bb", b"ccc", b"dddd")
        with WriteOpCtx() as write_op:
            self.ioctx.set_omap(write_op, keys, values)
            self.ioctx.operate_write_op(write_op, "test_obj")
        with ReadOpCtx() as read_op:
            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, keys)
            eq(ret, 0)
            self.ioctx.operate_read_op(read_op, "test_obj")
            eq(list(iter), list(zip(keys, values)))
        with WriteOpCtx() as write_op:
            self.ioctx.remove_omap_range2(write_op, "1", "4")
            self.ioctx.operate_write_op(write_op, "test_obj")
        with ReadOpCtx() as read_op:
            iter, ret = self.ioctx.get_omap_vals_by_keys(read_op, keys)
            eq(ret, 0)
            self.ioctx.operate_read_op(read_op, "test_obj")
            eq(list(iter), [("4", b"dddd")])

if __name__ == '__main__':
    unittest.main()
```

test results:
```
root@jw:/data/ceph/build# python3 test_range2.py 
.
----------------------------------------------------------------------
Ran 1 test in 2.803s

OK
```

Comment (tchaikov) [reply src/test/pybind/test_rados.py]: @zjcmszh thanks!

**Comment by tchaikov** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2020-12-03_06:25:29-rados-wip-kefu-testing-2020-12-02-2319-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2020-12-04_02:12:52-rados-wip-kefu-testing-2020-12-02-2319-distro-basic-smithi/

**Comment by zjcmszh** [comment]
@tchaikov thanks！

**Pull request merged**